### ISSUE TITLE
CT log worker stops unexpectedly - Failed to obtain bigger STH

### ISSUE THREAD

**Issue by d-Rickyy-b** [google/certificate-transparency-go#1019]
Hi there and happy new year to you,

every now and then I'm running into an issue where a worker routine for a specific ct log unexpectedly stops processing new logs.
Thanks to logging I managed to get the following entry:

```
[fetcher.go:195] https://ct.cloudflare.com/logs/nimbus2023: Failed to obtain bigger STH: Get "https://ct.cloudflare.com/logs/nimbus2023/ct/v1/get-sth?": read tcp x.x.x.x:36832->104.18.10.216:443: read: connection reset by peer
```

https://github.com/google/certificate-transparency-go/blob/5aefc51f2244711d9d5d68a80cdac120754bb29d/scanner/fetcher.go#L190-L200

On my side I am running this inside a go routine:
```go
scanErr := certScanner.Scan(ctx, w.foundCertCallback, w.foundPrecertCallback)
if scanErr != nil {
	log.Println("Scan error: ", scanErr)
	return
}
```

I did not get an error passed from `Scan()`, since the `fetcher.genRanges` method does not return the error that was received in there.
Is there an easy way to recover from this error on the library side? Or do I need to handle this on my side instead (e.g. by restarting the go routine)?

Regards,
Rico

**Comment by AlCutter** [comment]
Hi Rico,

Happy new year to you, too!
Thanks for the report and detailed investigation.

I think the main problem is that the go routine generating the ranges shouldn't just bail if there's an error fetching an updated STH when running in continuous mode. I've got a quick fix in #1022 which should prevent this particular issue happening again.

Cheers,
Al.

**Comment by AlCutter** [comment]
Closing this for now, do let us know if you have problems.
